pdns/pdns_recursor.cc

```diff
@@ -2897,50 +2929,94 @@ static void handleRunningTCPQuestion(int fd, FDMultiplexer::funcparam_t& var)
         ++g_stats.qcounter;
         ++g_stats.tcpqcounter;
 
+        if(dc->d_mdp.d_header.opcode == Opcode::Notify) {
+          if(!t_allowNotifyFrom || !t_allowNotifyFrom->match(dc->d_source)) {
+            if(!g_quiet) {
+              g_log<<Logger::Error<<"["<<MT->getTid()<<"] dropping TCP NOTIFY from "<<dc->d_source.toString()<<", address not matched by allow-notify-from"<<endl;
+            }
+
+            g_stats.unauthorizedNotify++;
+            return;
+          }
+
+          if(!isAllowNotifyForZone(qname)) {
+            if(!g_quiet) {
+              g_log<<Logger::Error<<"["<<MT->getTid()<<"] dropping TCP NOTIFY from "<<dc->d_source.toString()<<", for "<<qname.toLogString()<<", zone not matched by allow-notify-for"<<endl;
+            }
+
+            g_stats.disallowedNotify++;
+            return;
+          }
+        }
+
         string response;
         RecursorPacketCache::OptPBData pbData{boost::none};
 
-        /* It might seem like a good idea to skip the packet cache lookup if we know that the answer is not cacheable,
-           but it means that the hash would not be computed. If some script decides at a later time to mark back the answer
-           as cacheable we would cache it with a wrong tag, so better safe than sorry. */
-        dc->d_eventTrace.add(RecEventTrace::PCacheCheck);
-        bool cacheHit = checkForCacheHit(qnameParsed, dc->d_tag, conn->data, qname, qtype, qclass, g_now, response, dc->d_qhash, pbData, true, dc->d_source);
-        dc->d_eventTrace.add(RecEventTrace::PCacheCheck, cacheHit, false);
-
-        if (cacheHit) {
+        if(dc->d_mdp.d_header.opcode == Opcode::Query) {
+          /* It might seem like a good idea to skip the packet cache lookup if we know that the answer is not cacheable,
+             but it means that the hash would not be computed. If some script decides at a later time to mark back the answer
+             as cacheable we would cache it with a wrong tag, so better safe than sorry. */
+	  dc->d_eventTrace.add(RecEventTrace::PCacheCheck);
```

Comment:
indent issue

Reply:
This (and the other indent issues) display properly in Emacs and other tools on my system, but not in the GitHub viewer... and it's because the lines have mixed tabs/spaces for indentation. This is why I brought that topic up on the IRC channel yesterday... if we're OK with updating the .clang-format file to disable use of tabs for indentation, I'll reformat this branch.

Reply:
Fixed.